beets/dbcore/types.py

```diff
@@ -168,52 +209,66 @@ def format(self, value: int) -> str:
         return '{}{}'.format((value or 0) // self.unit, self.suffix)
 
 
-class Id(Integer):
+class Id(NullInteger):
     """An integer used as the row id or a foreign key in a SQLite table.
     This type is nullable: None values are not translated to zero.
     """
-    null = None
+    @property
+    def null(self) -> None:
+        return None
 
     def __init__(self, primary: bool = True):
         if primary:
             self.sql = 'INTEGER PRIMARY KEY'
 
 
-class Float(Type):
+class BaseFloat(Type[float, N]):
     """A basic floating-point type. The `digits` parameter specifies how
     many decimal places to use in the human-readable representation.
     """
     sql = 'REAL'
-    query = query.NumericQuery
+    query = NumericQuery
     model_type = float
 
     def __init__(self, digits: int = 1):
         self.digits = digits
 
-    def format(self, value: float) -> str:
+    def format(self, value: Union[float, N]) -> str:
         return '{0:.{1}f}'.format(value or 0, self.digits)
 
 
-class NullFloat(Float):
+class Float(BaseFloat[float]):
+    pass
+
+
+class NullFloat(BaseFloat[None]):
     """Same as `Float`, but does not normalize `None` to `0.0`.
     """
-    null = None
+    @property
+    def null(self) -> None:
+        return None
 
 
-class String(Type):
+class BaseString(Type[T, N]):
     """A Unicode string type.
     """
     sql = 'TEXT'
-    query = query.SubstringQuery
+    query = SubstringQuery
 
-    def normalize(self, value: str) -> str:
+    def normalize(self, value: N) -> Union[T, N]:
```

Comment:
Perhaps `value` should have type `Any`, as it does elsewhere?

Reply:
Good catch, thanks!